src/coreclr/jit/morph.cpp

```diff
@@ -11144,6 +11212,11 @@ GenTree* Compiler::fgOptimizeCastOnAssignment(GenTreeOp* asg)
 
     if (genActualType(castFromType) == genActualType(castToType))
     {
+        // If the cast is marked as do-not-cse, then
+        // we should not remove it to imply an implicit cast.
+        if (cast->gtFlags & GTF_DONT_CSE)
+            return asg;
```

Comment:
This worries me. The fact we see the VN bug exposed with this transformation, it is risky to take the transformation without fixing the VN bug first.

Reply:
I agree, which is why lowering was on the table since it is less risky.